cookbooks/equilibrium_grain_size/plugins/equilibrium_grain_size.cc

```diff
@@ -0,0 +1,1750 @@
+        {
+          if (material_file_format == perplex)
+            material_lookup.push_back(std::shared_ptr<MaterialUtilities::Lookup::MaterialLookup>
+                                      (new MaterialUtilities::Lookup::PerplexReader(datadirectory+material_file_names[i],
+                                                                                    /*use_bilinear_interpolation*/ true,
+                                                                                    this->get_mpi_communicator())));
+          else if (material_file_format == hefesto)
+            material_lookup.push_back(std::shared_ptr<MaterialUtilities::Lookup::MaterialLookup>
+                                      (new MaterialUtilities::Lookup::HeFESToReader(datadirectory+material_file_names[i],
+                                                                                    datadirectory+derivatives_file_names[i],
+                                                                                    /*use_bilinear_interpolation*/ true,
+                                                                                    this->get_mpi_communicator())));
+          else
+            AssertThrow (false, ExcNotImplemented());
+        }
+    }
+
+
+
+    template <int dim>
+    void
+    EquilibriumGrainSize<dim>::update()
+    {
+      if (use_depth_dependent_viscosity)
+        {
+          std::vector<std::unique_ptr<internal::FunctorBase<dim>>> lateral_averaging_properties;
+          lateral_averaging_properties.push_back(std::make_unique<internal::FunctorDepthAverageUnscaledViscosity<dim>>());
+
+          std::vector<std::vector<double>> averages =
+            this->get_lateral_averaging().compute_lateral_averages(reference_viscosity_coordinates,
+                                                                   lateral_averaging_properties);
+
+          average_viscosity_profile.swap(averages[0]);
+
+          for (const auto &lateral_viscosity_average: average_viscosity_profile)
+            AssertThrow(numbers::is_finite(lateral_viscosity_average),
+                        ExcMessage("In computing depth averages, there is at"
+                                   " least one depth band that does not have"
+                                   " any quadrature points in it."
+                                   " Consider reducing number of depth layers"
+                                   " for averaging."));
+        }
+
+      initialized = true;
+    }
+
+
+
+    template <int dim>
+    std::pair<double, unsigned int>
+    EquilibriumGrainSize<dim>::get_reference_viscosity (const double depth) const
+    {
+      // Make maximal depth slightly larger to ensure depth < maximal_depth
+      const double maximal_depth = this->get_geometry_model().maximal_depth() *
+                                   (1.0+std::numeric_limits<double>::epsilon());
+      (void) maximal_depth;
+
+      Assert(depth < maximal_depth, ExcInternalError());
```

Comment:
Would you also want the corresponding minimum check?
```suggestion
      Assert(depth < maximal_depth, ExcInternalError());
      Assert(depth > -maximal_depth*std::numeric_limits<double>::epsilon(), ExcInternalError());
```

Reply:
I don't remember why we are doing this check, but I think that suggestion makes sense.